python/tk_desktop/desktop_window.py

```diff
@@ -738,31 +785,78 @@ def _get_toolkit_classic_pipeline_configurations(self, connection, project):
-        pipeline_configurations = connection.find(
+        found_pcs = connection.find(
             "PipelineConfiguration",
             [["project", "is", project]],
-            fields=["mac_path", "windows_path", "linux_path", "users", "code"]
-        )
-        # Ideally we would filter out in the Shotgun query any entry that has a plugin id set.
-        # Unfortunately, at the time of writing plugin ids are not live on nost Shotgun instances.
-        # So instead of filtering everything in the query, we'll filter out entries locally with the
-        # following.
-        pipeline_configurations = filter(
-            lambda pc: not("sg_plugin_ids" in pc or "plugin_ids" in pc), pipeline_configurations
+            fields=["mac_path", "windows_path", "linux_path", "users", "code", "sg_plugin_ids", "plugin_ids", "project"]
         )
 
-        for pc in pipeline_configurations:
+        def is_pipeline_usable(pc):
+            """
+            Ensures that a pipeline is not zero-config based and that it is accessible for the current user.
+            """
+            # If there is a plugin id, can't use this.
+            if pc.get("sg_plugin_ids") or pc.get("plugin_ids"):
+                return False
+
+            if not ShotgunPath.from_shotgun_dict(pc):
+                log.warning("Skipping Toolkit Classic pipeline configuration id %s without any path set." % pc["id"])
+                return False
+
+            # If there are no users assigned, this pipeline is accessible from for everyone.
+            if not pc["users"]:
+                return True
+
+            # Search for ourselves.
+            for user in pc["users"]:
+                # We've found ourselves, awesome!
+                if self._current_user_id == user["id"]:
+                    return True
+
+            return False
+
+        # FIXME: We need to discuss what we do with multiple primaries with Toolkit classic. In zero config we keep
+        # the earliest one. It follows that the new desktop should follow the same rules, but the behavior already
+        # exists in the wild, so we may have to keep supporting that one.
+
+        # Filter out pipelines that can't be accessed.
+        accessible_pcs = filter(is_pipeline_usable, found_pcs)
+
+        # Massage the dictionaries keys so it uses the same key names of the zero config pipelines.
+        for pc in accessible_pcs:
             pc["name"] = pc["code"]
             del pc["code"]
 
-        return pipeline_configurations
+        return accessible_pcs
+
+    def _is_primary_pc(self, pc):
+        """
+        Tests if a pipeline configuration is a primary.
+
+        :param pc: Pipeline configuration entity with key ``code``.
+
+        :returns: True if the pipeline configuration is a primary, else otherwise.
+        """
+        return pc["name"] == constants.PRIMARY_PIPELINE_CONFIG_NAME
+
+    def _merge_pipeline_configuration_lists(self, classic_pipelines, zero_config_pipelines):
```

Comment:
how about we say `bootstrap_pipeline_configs` instead of zero config. Zero config for me is just one possible configuration, but they all come from the bootstrap API.